### ISSUE TITLE
Support dynamic anchor names, depending on value inside struct?

### ISSUE THREAD

**Issue by timgws** [goccy/go-yaml#69]
This is going to be a little tricky to describe, but hopefully with the attached code you should be able to see what I am trying to do.

I want to make a YAML document that looks something like:
```yaml
hosts:
- host: &host1
    hostname: host1.example.com
    username: queue1
    password: queue1
- host: &host2
    hostname: host2.example.com
    username: queue1
    password: queue1
queues:
- name: queue
  host: *host1
```

This is the code that I have written at the moment:

```golang
import (
	"fmt"
	"github.com/goccy/go-yaml"
)

type HostList struct {
	Host *Host `yaml:",inline,anchor"`
	Name     string `yaml:","`
}

type Host struct {
	*Host `yaml:",omitempty"`
	Name string `yaml:",inline,anchorname"`
	Hostname string
	Username string
	Password string
}

type Queue struct {
	Name     string `yaml:","`
	*Host    `yaml:",alias"`
}

func main() {
	var doc struct {
		HostList []*HostList `yaml:"hosts"`
		Hosts    []*Host     `yaml:"hosts2"`
		Queues   []*Queue    `yaml:"queues"`
	}
	host1 := &Host {
			Hostname: "host1.example.com",
			Username: "queue1",
			Password: "queue1",
	}
	host2 := &Host {
			Hostname: "host2.example.com",
			Username: "queue1",
			Password: "queue1",
	}

	doc.HostList = []*HostList{
		{
			Name: "host1",
			Host: host1,
		},{
			Name: "host2",
			Host: host2,
		},
	}
	doc.Queues = []*Queue {
		{
			Name: "queue",
			Host: host1,
		},{
			Name: "queue2",
			Host: host2,
		},
	}

	bytes, err := yaml.Marshal(doc)
	if err != nil {
		fmt.Println(err)
		return
	}

	fmt.Printf("%s", string(bytes))
}
```

I can't seem to work out how to get the names out of the ast.

https://github.com/goccy/go-yaml/blob/009880f5ba81cb7637bce48b0e5ea90a49238725/encode.go#L394-L403

Can I please get some help?

**Comment by goccy** [comment]
Hi @timgws 

I tried adding a feature to solve this problem in # 70. 
Can you do what you want with this  ?

**Comment by timgws** [comment]
> Can you do what you want with this ?

Looking at the test that you have written, yes!

**Comment by timgws** [comment]
One issue now is that I can not unmarshal this document.

```go
yaml_file := `
hosts:
- host: &host1
    hostname: host1.example.com
    username: userA
    password: pass1
- host: &host2
    hostname: host2.example.com
    username: userB
    password: pass2
queues:
- name: queue
  host: *host1
- name: queue2
  host: *host2
- name: queue3
  host: *host2
`


	var doc2 struct {
		Hosts  []*HostDecl `yaml:"hosts"`
		Queues []*Queue    `yaml:"queues"`
	}

	err := yaml.Unmarshal([]byte(yaml_file), &doc2)
	fmt.Println(err)
	fmt.Println(doc2.Queues[0].Host.Hostname)
	fmt.Println(doc2.Queues[1].Host.Hostname)
	fmt.Println(doc2.Queues[2].Host.Hostname)
```

I get back the error:
```
cannot unmarshal map[string]interface {} into Go struct field .Hosts of type string
```

**Comment by goccy** [comment]
I released as v1.1.9 ( https://github.com/goccy/go-yaml/releases/tag/v1.1.9 )
